### PR TITLE
test(react): add snapshot test for public API

### PR DESCRIPTION
**Pull request by joshblack**
Adds in a test that snapshots our public API for `carbon-components-react`. In particular, this test uses our `src/index.js` entrypoint to determine which components we export and defines their public API through a combination of the components' `propTypes` and other properties available. To get prop type information, we mock the `prop-types` package to return the type information.

#### Changelog

**New**

- Add a `PublicAPI-test.js` file that determines the public API of our components and creates a snapshot of these values

**Changed**

**Removed**

#### Testing / Reviewing

- Verify the public API snapshot isn't crazy 😅

### CODE REVIEW

**Review by abbeyhrt — APPROVED** [review]
This looks good to me! Out of curiosity, would this slow anything down? Like our build time or something?

**Comment by joshblack** [comment]
@abbeyhrt nope! Should only be an extra test 😄

File: packages/react/__tests__/PublicAPI-test.js [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    const Component = CarbonReact[name];
+    PublicAPI.set(name, mapComponentToAPI(Component));
+  });
+
+  function mapComponentToAPI(Component) {
+    const api = {};
+
+    Object.keys(Component).forEach(key => {
+      // There are a couple of properties on components that we don't believe
+      // are part of our API, such `_` prefixed variables, or capture details
```
Comment (joshblack) [inline packages/react/__tests__/PublicAPI-test.js]:
```suggestion
      // are part of our API, such as `_` prefixed variables, or capture details
```

**Review by tw15egan — APPROVED** [review]
I'm all for it 👍 Changes made to `PropTypes` should be intentional,  and this is a nice way to catch any unintended changes

File: CODEOWNERS [hunk 2]
```diff
@@ -10,3 +10,8 @@
 
 # Brand
 /packages/colors/artifacts/ @lisalind-ibm
+
+# Release team should be notified of Public API changes in the following
+# packages
+/packages/react/__tests__/PublicAPI-test.js @carbon-design-system/developers-sytem @carbon-design-system/release
+/packages/react/__tests__/__snapshots__/PublicAPI-test.js.snap @carbon-design-system/developers-sytem @carbon-design-system/release
```
Comment (joshblack) [inline CODEOWNERS]:
```suggestion
/packages/react/__tests__/PublicAPI-test.js @carbon-design-system/developers-system @carbon-design-system/release
/packages/react/__tests__/__snapshots__/PublicAPI-test.js.snap @carbon-design-system/developers-system @carbon-design-system/release
```

**Pull request merged**